### PR TITLE
Trainer - deprecate tokenizer for processing_class

### PR DESCRIPTION
**Pull request by amyeroberts**
For reviewers: most files touched here are just updates to the documentation. Unfortunately, a lot of the diff is just stripping of unnecessary whitespaces (which my editor does automatically). 

The most important changes are in: 
* src/transformers/trainer.py
* src/transformers/trainer_callback.py
* src/transformers/trainer_seq2seq.py

-------------------------------------------

# What does this PR do?

At the moment, if we wish for a processing class e.g. an image processor, to be saved alongside the model e.g. when pushing to the hub, we have to do the following: 

```py
trainer = Trainer(
    model, 
    args, 
    tokenizer=image_processor, 
    ...
)
```

This causes a lot of confusion for users (an image processor is not a tokenizer). 

Previous efforts were made to add individual classes to the trainer s.t. one could do: `trainer = Trainer(model, args, image_processor=image_processor, ...)`, however, this has some drawbacks: 
* It creates ambiguity when a model has a processor class: should we be able to do `Trainer(model, args, processor=processor)` or should it be `Trainer(model, args, tokenizer=tokenizer, image_processor=image_processor)`? 
* It adds new arguments to handle for every processing class, meaning more work if future classes are added

This PR therefore just deprecates `tokenizer` in favour of a more general `processing_class` argument, which can be any class with a `from_pretrained` method. 

Relevant PRs and discussions: 
* #29896
* #30129 
* #30102 
* #30864

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by SangbumChoi** [comment]
Love this (Always sad about naming writing `tokenizer=image_processor`)

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -60,7 +60,9 @@
 from .data.data_collator import DataCollator, DataCollatorWithPadding, default_data_collator
 from .debug_utils import DebugOption, DebugUnderflowOverflow
 from .feature_extraction_sequence_utils import SequenceFeatureExtractor
+from .feature_extraction_utils import FeatureExtractionMixin
```
Comment (amyeroberts) [inline src/transformers/trainer.py:63]: This file contains the most important logic changes for review

File: src/transformers/trainer_callback.py [hunk 2]
```diff
@@ -397,12 +399,12 @@ def on_prediction_step(self, args: TrainingArguments, state: TrainerState, contr
 class CallbackHandler(TrainerCallback):
     """Internal class that just calls the list of callbacks in order."""
 
-    def __init__(self, callbacks, model, tokenizer, optimizer, lr_scheduler):
+    def __init__(self, callbacks, model, processing_class, optimizer, lr_scheduler):
```
Comment (amyeroberts) [inline src/transformers/trainer_callback.py:402]: I directly just swapped the names here, as the docstring says the object is just an internal class

**Comment by amyeroberts** [comment]
@molbap If you have capacity - do you think you could do a first review of this? Arthur and Zack are super busy so would be good to have someone who knows the processors well to review in their stead

**Review by molbap — APPROVED** [review]
Checked out the rework, nice - saw that now you can't pass both `tokenizer` and `processing_class`, it will rightfully raise an error, much better. lgtm!

**Comment by amyeroberts** [comment]
Asking for review from @SunMarc for Trainer as Zach is off. As this is a very public API I'd like to get a few eyes on it, especially with people who know it well to make sure it's OK!

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -421,6 +433,14 @@ def __init__(
         # force device and distributed setup init explicitly
         args._setup_devices
 
+        if tokenizer is not None:
+            if processing_class is not None:
+                raise ValueError(
+                    "You cannot specify both `tokenizer` and `processing_class` at the same time. Please use `processing_class`."
+                )
+            logger.warning("`tokenizer` is now deprecated, please use `processing_class` instead.")
```
Comment (ArthurZucker) [inline src/transformers/trainer.py:441]: do we want futur warning?

**Review by ArthurZucker — APPROVED** [review]
Clean, nothing to say! Thanks for ths!

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -3741,6 +3744,98 @@ def test_eval_use_gather_object(self):
         _ = trainer.evaluate()
         _ = trainer.predict(eval_dataset)
 
+    def test_trainer_saves_tokenizer(self):
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py:3747]: Thanks for adding these nice tests !

File: src/transformers/trainer.py (hunk 3 above)
Comment (SunMarc) [reply src/transformers/trainer.py:441]: Maybe say that it will be removed in the v5 ?

**Review by SunMarc — APPROVED** [review]
Thanks for the PR @amyeroberts ! LGTM !

File: src/transformers/trainer.py (hunk 3 above)
Comment (amyeroberts) [reply src/transformers/trainer.py:441]: I'll move it to a future warning and specify v5 removal :)

**Comment by goea-shuhei** [comment]
_pad_tensors_to_max_len in Seq2SeqTrainer still uses tokenizer instead of  processing_class

**Pull request merged**